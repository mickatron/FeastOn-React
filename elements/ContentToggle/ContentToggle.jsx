import PropTypes from 'prop-types';
import React from 'react';
import UIWrapperHOC from '../../HOC/uiWrapper/uiWrapper';

import keyType from '../../proptypes/key';

class ContentToggle extends React.PureComponent {
  static propTypes = {
    keyId: keyType,
    /** Disable the ContentToggle. */
    disabled: PropTypes.bool,
    /** ContentToggle header props. */
    header: PropTypes.func.isRequired,
    /** ContentToggle content props. */
    content: PropTypes.func.isRequired,
    /** Is the ContentToggle open. */
    open: PropTypes.bool,

    onToggle: PropTypes.func,
  }

  static defaultProps = {
    keyId: undefined,
    disabled: false,
    onToggle: undefined,
    open: false,
  }

  static getDerivedStateFromProps(nextProps, state) {
    if (nextProps.open !== state.open) {
      return { open: nextProps.open };
    }
    return state;
  }

  state = {
    open: this.props.open,
  }

  onToggleHandler = (event) => {
    if (this.props.disabled) return;
    const { keyId, onToggle } = this.props;
    const { open } = this.state;
    this.setState({ open: !open }, () => {
      if (onToggle) onToggle(event, { keyId, open: this.state.open });
    });
  }

  render() {
    const {
      content,
      header,
    } = this.props;
    const { open } = this.state;
    return (
      <React.Fragment>
        {header({ onToggle: this.onToggleHandler, open })}
        {open && content({ open })}
      </React.Fragment>
    );
  }
}

export default UIWrapperHOC({
  name: 'ContentToggle',
  Component: ContentToggle,
  modifiers: ['open', 'disabled'],
});
