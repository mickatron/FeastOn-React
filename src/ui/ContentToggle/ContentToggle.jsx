import ChevronIcon from 'react-material-icon-svg/dist/ChevronLeftIcon';

import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import classNameType from 'js/shared/proptypes/className';
import Button from 'js/shared/components/ui/Button/Button';

import normalizeEventName from 'js/shared/util/normalizeEventName';

export default class ContentToggle extends React.PureComponent {

  classPrefix = 'c-ContentToggle'; 
  static propTypes = {
    children: PropTypes.node,
    className: classNameType,
    disabled: PropTypes.bool,
    header: PropTypes.node.isRequired,
    mode: PropTypes.oneOf([
      'accordion',
      'dropDown',
    ]),
    open: PropTypes.bool,
    Icon: PropTypes.oneOfType([
      PropTypes.element,
      PropTypes.instanceOf(null),
    ]),
    iconPos: PropTypes.oneOf([ 'right', 'left' ]),
    triggerEvent:PropTypes.oneOf([
      'onClick',
      'onMouseEnter',
    ]),
    exitEvent: PropTypes.oneOf([
      'onClick',
      'onMouseLeave',
    ]),
  }

  static defaultProps = {
    disabled: false,
    Icon: ChevronIcon,
    iconPos: 'right',
    mode: 'accordion',
    open: false,
    triggerEvent: 'onClick', 
    exitEvent: 'onClick',
  }

  state = {
    open: this.props.open,
  }

  onClickHandler = (event) => {
    if (this.props.disabled) return;
    const open = this.state.open;
    const type = event.type;

    if(
      (!open && type === normalizeEventName(this.props.triggerEvent))
      || 
      (open && type === normalizeEventName(this.props.exitEvent))
    ) {
      this.setState({ open: !open });
    }
    return;
  }

  render() {
    const {
      children,
      className,
      disabled,
      header,
      Icon,
      iconPos,
      mode,
      exitEvent,
      triggerEvent,
    } = this.props;

    const extraClassNames = {
      [`${this.classPrefix}--open`]: this.state.open,
      [`${this.classPrefix}--disabled`] : disabled,
    };

    return (
      <div 
        className={
          classnames(
            className, 
            this.classPrefix, 
            `${this.classPrefix}--${mode}`, 
            extraClassNames,
          )}
        {...{ [triggerEvent]: this.onClickHandler }}
        {...{ [exitEvent]: this.onClickHandler }}
      >
        <Button 
          appearance="link" 
          className={`${this.classPrefix}__header`} 
        >
          {iconPos === 'left' && Icon ? <Icon className={`${this.classPrefix}__icon`} /> : null }
          {header}
          {iconPos === 'right' && Icon ? <Icon className={`${this.classPrefix}__icon`} /> : null }
        </Button>
        <div className={`${this.classPrefix}__content`}>
          {children}
        </div>
      </div>
    );
  }
}