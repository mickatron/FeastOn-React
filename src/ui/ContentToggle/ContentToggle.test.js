import { shallow } from 'enzyme';
import React from 'react';
import ContentToggle from './ContentToggle';

describe('ContentToggle', () => {

  let component = null;
  describe('renders', () => {
    it('ContentToggle', () => {
      component = shallow(
        <ContentToggle
          className="test"
          header="some test header value"
          open={true}
        >
          <p>Some ContentToggle test text.</p>
        </ContentToggle>
      );
      expect(component).toMatchSnapshot();
      component.find('.c-ContentToggle__header').simulate('click', {});
      expect(component.state('open')).toBe(false);
    });

    it('closed & disabled ContentToggle', () => {
      component = shallow(
        <ContentToggle 
          className="test"
          header="some test header value"
          open={false}
          disabled={true}
        >
          <p>Some ContentToggle test text.</p>
        </ContentToggle>
      );
      expect(component).toMatchSnapshot();
      component.find('.c-ContentToggle__header').simulate('click', {});
      expect(component.state('open')).toBe(false);
    });

  });

});