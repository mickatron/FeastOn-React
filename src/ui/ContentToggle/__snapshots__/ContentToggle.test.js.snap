// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContentToggle renders ContentToggle 1`] = `
<div
  className="test c-ContentToggle c-ContentToggle--accordion c-ContentToggle--open"
  onClick={[Function]}
>
  <Button
    appearance="link"
    className="c-ContentToggle__header"
    iconPos="left"
    mode="button"
    size="delta"
  >
    some test header value
    <ChevronLeftIcon
      className="c-ContentToggle__icon"
    />
  </Button>
  <div
    className="c-ContentToggle__content"
  >
    <p>
      Some ContentToggle test text.
    </p>
  </div>
</div>
`;

exports[`ContentToggle renders closed & disabled ContentToggle 1`] = `
<div
  className="test c-ContentToggle c-ContentToggle--accordion c-ContentToggle--disabled"
  onClick={[Function]}
>
  <Button
    appearance="link"
    className="c-ContentToggle__header"
    iconPos="left"
    mode="button"
    size="delta"
  >
    some test header value
    <ChevronLeftIcon
      className="c-ContentToggle__icon"
    />
  </Button>
  <div
    className="c-ContentToggle__content"
  >
    <p>
      Some ContentToggle test text.
    </p>
  </div>
</div>
`;
